### PR TITLE
[DSM] DDP-7556_Mercury_fixes

### PR DESCRIPTION
**Pull request by pegahtah**
PR comments applied and changes to the request to also have collaborator participant id

### CODE REVIEW

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/ddp/instance/DDPInstanceDao.java [hunk 1]
```diff
@@ -42,18 +42,21 @@ public class DDPInstanceDao implements Dao<DDPInstanceDto> {
...
                     + "notification_recipients = ?, migrated_ddp = ?, billing_reference = ?, es_participant_index = ?,"
-                    + "es_activity_definition_index = ?, es_users_index = ?, study_pre_filter = ?, research_project = ?";
+                    + "es_activity_definition_index = ?, es_users_index = ?, study_pre_filter = ?, research_project = ?, "
+                    + "mercury_order_creator= ? ";
 
     private static final String SQL_DELETE_DDP_INSTANCE = "DELETE FROM ddp_instance WHERE ddp_instance_id = ?";
     private static final String SQL_SELECT_INSTANCE_WITH_ROLE =
             "SELECT ddp_instance_id, instance_name, base_url, collaborator_id_prefix, migrated_ddp, billing_reference, "
-                    + "es_participant_index, es_activity_definition_index, es_users_index, study_pre_filter, research_project, "
+                    +
```
Comment (SimoneMaiwald) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/ddp/instance/DDPInstanceDao.java]: can we not do just `+` in a line… move your `mercury….` into the same line as `(select`, please

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/pubsub/MercuryOrderPublisher.java [hunk 2]
```diff
@@ -98,44 +96,28 @@ public void onSuccess(String messageId) {
...
         log.info("Publishing message to mercury");
         String researchProject = ddpInstance.getResearchProject();
-        String creatorId = new InstanceSettings().getInstanceSettings(ddpInstance.getName()).getMercuryOrderCreator().orElseThrow();
+        String creatorId = ddpInstance.getMercuryOrderCreator().orElseThrow();
         String mercuryOrderId = createMercuryUniqueOrderId();
         MercuryPdoOrder mercuryPdoOrder = new MercuryPdoOrder(creatorId, mercuryOrderId, researchProject, barcodes);
         String json = new Gson().toJson(mercuryPdoOrder);
         if (StringUtils.isNotBlank(json)) {
-            orderOnMercury(projectId, topicId, json, barcodes, ddpParticipantId, mercuryOrderId);
+            this.mercuryOrderDao.orderOnMercury(projectId, topicId, json, barcodes, ddpParticipantId, mercuryOrderId, this);
```
Comment (SimoneMaiwald) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/pubsub/MercuryOrderPublisher.java]: I don't like to give in `this` to `orderOnMercury` please refactor to have the dao method return what is needed and then order it from here!

**Comment by SimoneMaiwald** [comment]
please also remove `mercury_order_creator` from `instance_settings` on dev if you haven't done so yet

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/ddp/instance/DDPInstanceDao.java (hunk 1 above)
Comment (pegahtah) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/ddp/instance/DDPInstanceDao.java]: ugh I missed this , I'll fix

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/pubsub/MercuryOrderPublisher.java (hunk 2 above)
Comment (pegahtah) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/pubsub/MercuryOrderPublisher.java]: So if you look at the `OrderOnMercury` I am doing the insertion of db and the publish on pub sub in thee same method to be  able to roll back the db when sth goes wrong with pubsub.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/pubsub/MercuryOrderPublisher.java [hunk 3]
```diff
@@ -98,44 +95,34 @@ public void onSuccess(String messageId) {
...
-            return result;
-        });
-        if (results.resultException != null) {
-            throw new RuntimeException("Error inserting new order ", results.resultException);
+
         }
+
     }
+
+    private boolean requestHadCollaboratorId(String ddpParticipantId, String collaboratorParticipantId) {
```
Comment (SimoneMaiwald) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/pubsub/MercuryOrderPublisher.java]: is that just for testing purpose? we should know what they send in the message

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/mercury/PostMercuryOrderDummyRoute.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import org.broadinstitute.dsm.statics.UserErrorMessages;
+import org.broadinstitute.dsm.util.UserUtil;
+import org.broadinstitute.lddp.handlers.util.Result;
+import spark.QueryParamsMap;
+import spark.Request;
+import spark.Response;
+import spark.Route;
+
+@Slf4j
+public class PostMercuryOrderDummyRoute implements Route {
```
Comment (pegahtah) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/mercury/PostMercuryOrderDummyRoute.java:21]: renamed route, added `Dummy` in it

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/statics/DBConstants.java [hunk 5]
```diff
@@ -512,7 +513,5 @@ public class DBConstants {
     public static final String COHORT_TAG = "cohort_tag";
     public static final String COHORT_TAG_NAME = "cohort_tag_name";
 
-    //instance_settings
-    public static final String MERCURY_ORDER_CREATOR = "mercury_order_creator";
```
Comment (pegahtah) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/statics/DBConstants.java:522]: moved to ddp_instance

File: pepper-apis/dsm-core/src/main/resources/liquibase/CMI/DDP-7556_mercury_sequencing.xml [hunk 6]
```diff
@@ -26,12 +26,20 @@
...
             <column name="research_project" type="VARCHAR(200)"></column>
         </addColumn>
-        <addColumn tableName="instance_settings">
+        <addColumn tableName="ddp_instance">
             <column name="mercury_order_creator" type="VARCHAR(200)"></column>
         </addColumn>
     </changeSet>
+    <changeSet id="DDP-7556_additional_mercury_sequencing" author="pegah">
+        <addColumn tableName="mercury_sequencing">
+            <column name="created_by" type="VARCHAR(200)"></column>
```
Comment (pegahtah) [inline pepper-apis/dsm-core/src/main/resources/liquibase/CMI/DDP-7556_mercury_sequencing.xml:39]: user who submits the order

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/CreateClinicalDummyKitRoute.java [hunk 7]
```diff
@@ -96,13 +95,20 @@ public Object handle(Request request, Response response) {
             ddpParticipantId = new BSPDummyKitDao().getRandomParticipantForStudy(ddpInstance);
             maybeParticipantByParticipantId =
                     ElasticSearchUtil.getParticipantESDataByParticipantId(ddpInstance.getParticipantIndexES(), ddpParticipantId);
-            if (maybeParticipantByParticipantId.isEmpty()) {
-                throw new RuntimeException("PT not found " + ddpParticipantId);
+            while (tries < 10 && (maybeParticipantByParticipantId.isEmpty() || !participantIsEnrolled(maybeParticipantByParticipantId)
+                    || maybeParticipantByParticipantId.get().getProfile().map(ESProfile::getHruid).isEmpty())) {
```
Comment (pegahtah) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/CreateClinicalDummyKitRoute.java:99]: I moved the participant enrolled check into this method

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/pubsub/MercuryOrderPublisher.java [hunk 8]
```diff
@@ -98,44 +101,31 @@ public void onSuccess(String messageId) {
...
+        String ddpParticipantId = maybeParticipantId.orElseThrow();
         log.info("Publishing message to mercury");
-        String researchProject = ddpInstance.getResearchProject();
-        String creatorId = new InstanceSettings().getInstanceSettings(ddpInstance.getName()).getMercuryOrderCreator().orElseThrow();
+        String researchProject = ddpInstance.getResearchProject().orElseThrow();
+        String creatorId = ddpInstance.getMercuryOrderCreator().orElseThrow();
         String mercuryOrderId = createMercuryUniqueOrderId();
         MercuryPdoOrder mercuryPdoOrder = new MercuryPdoOrder(creatorId, mercuryOrderId, researchProject, barcodes);
-        String json = new Gson().toJson(mercuryPdoOrder);
+        MercuryPdoOrderBase mercuryPdoOrderBase = new MercuryPdoOrderBase(mercuryPdoOrder);
```
Comment (pegahtah) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/pubsub/MercuryOrderPublisher.java:115]: changes in Json needed all the order going in to one `productOrder` object

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/pubsub/MercuryOrderPublisher.java [hunk 9]
```diff
@@ -98,44 +101,31 @@ public void onSuccess(String messageId) {
...
-            SimpleResult result = new SimpleResult();
             try {
-                for (MercuryOrderDto order : newOrders) {
-                    result = this.mercuryOrderDao.create(order, conn);
-                    if (result.resultException != null) {
-                        return result;
-                    }
-                }
-                publishWithErrorHandler(projectId, topicId, json);
+                List<MercuryOrderDto> newOrders = MercuryOrderUseCase.createAllOrders(barcodes, ddpParticipantId, mercuryOrderId, userId);
```
Comment (pegahtah) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/pubsub/MercuryOrderPublisher.java:119]: Creating all the `MercuryOrderDto` objects for submitted barcodes

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/pubsub/MercuryOrderPublisher.java [hunk 10]
```diff
@@ -98,44 +101,31 @@ public void onSuccess(String messageId) {
...
+                this.publishWithErrorHandler(projectId, topicId, json);
+                this.mercuryOrderDao.insertMercuryOrders(newOrders);
```
Comment (pegahtah) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/pubsub/MercuryOrderPublisher.java:121]: only inserting if the pubsub message was successfully published

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/mercury/MercuryOrderDao.java [hunk 11]
```diff
@@ -17,43 +19,49 @@
...
+    public HashMap<String, MercuryOrderDto> getPossibleBarcodesForParticipant(String ddpParticipantId) {
+        HashMap<String, MercuryOrderDto> map = new HashMap<>();
         SimpleResult results = inTransaction((conn) -> {
             SimpleResult dbVals = new SimpleResult();
-            try (PreparedStatement stmt = conn.prepareStatement(SQL_GET_KIT_FROM_BARCODE)) {
-                stmt.setString(1, barcode);
-                stmt.setString(2, barcode);
-                stmt.setString(3, ddpParticipantId);
+            try (PreparedStatement stmt = conn.prepareStatement(SQL_GET_KIT_FROM_BARCODE_KIT_LABEL)) {
+                stmt.setString(1, ddpParticipantId);
```
Comment (pegahtah) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/mercury/MercuryOrderDao.java:54]: finding all eligible kit labels for participant

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/mercury/MercuryOrderDao.java [hunk 12]
```diff
@@ -63,9 +71,33 @@ public Optional<MercuryOrderDto> getMercuryOrderFromSMIdOrKitLabel(String barcod
...
-            throw new RuntimeException("Error getting info for barcode " + barcode, results.resultException);
+            throw new RuntimeException("Error getting possible mercury orders for participant " + ddpParticipantId,
+                    results.resultException);
         }
-        return Optional.ofNullable((MercuryOrderDto) results.resultValue);
+
+        results = inTransaction((conn) -> {
+            SimpleResult dbVals = new SimpleResult();
+            try (PreparedStatement stmt = conn.prepareStatement(SQL_GET_KIT_FROM_BARCODE_SM_ID)) {
+                stmt.setString(1, ddpParticipantId);
```
Comment (pegahtah) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/mercury/MercuryOrderDao.java:81]: finding all eligible sm ids for participant

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dto/mercury/MercuryOrderUseCase.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+                throw new NoSuchElementException(message);
+            }
+            MercuryOrderDto order = participantBarcodes.get(barcode);
+            order.setOrderId(orderId);
+            order.setOrderDate(System.currentTimeMillis());
+            order.setCreatedBy(userId);
+            orders.add(order);
+        });
+        return orders;
+    }
```
Comment (pegahtah) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dto/mercury/MercuryOrderUseCase.java:31]: Finding all the kits and sm ids that are eligible for ordering for that participant in here, choosing the ones that were in the order and adding them to the orders object.

If a label was not eligible, throwing an error

**Review by gmakharat — APPROVED** [review]

**Pull request merged**